syntax/parser_arithm.go

```diff
@@ -0,0 +1,482 @@
+	value := p.arithmExprBand(compact)
+	for BinAritOperator(p.tok) == Xor {
+		if compact && p.spaced {
+			return value
+		}
+		op := p.tok
+		pos := p.pos
+		p.nextArithOp(compact)
+		y := p.arithmExprBand(compact)
+		if y == nil {
+			p.followErrExp(pos, op.String())
+		}
+		value = &BinaryArithm{
+			OpPos: pos,
+			Op:    BinAritOperator(op),
+			X:     value,
+			Y:     y,
+		}
+	}
+	return value
+}
+
+func (p *Parser) arithmExprBand(compact bool) ArithmExpr {
+	value := p.arithmExpr5(compact)
+	for BinAritOperator(p.tok) == And {
+		if compact && p.spaced {
+			return value
+		}
+		op := p.tok
+		pos := p.pos
+		p.nextArithOp(compact)
+		y := p.arithmExpr5(compact)
+		if y == nil {
+			p.followErrExp(pos, op.String())
+		}
+		value = &BinaryArithm{
+			OpPos: pos,
+			Op:    BinAritOperator(op),
+			X:     value,
+			Y:     y,
+		}
+	}
+	return value
+}
+
+func (p *Parser) arithmExpr5(compact bool) ArithmExpr {
+	value := p.arithmExpr4(compact)
+	for BinAritOperator(p.tok) == Eql || BinAritOperator(p.tok) == Neq {
+		if compact && p.spaced {
+			return value
+		}
+		op := p.tok
+		pos := p.pos
+		p.nextArithOp(compact)
+		y := p.arithmExpr4(compact)
+		if y == nil {
+			p.followErrExp(pos, op.String())
+		}
+		value = &BinaryArithm{
+			OpPos: pos,
+			Op:    BinAritOperator(op),
+			X:     value,
+			Y:     y,
+		}
+	}
+	return value
+}
+
+func (p *Parser) arithmExpr4(compact bool) ArithmExpr {
+	value := p.arithmExprShift(compact)
+	for BinAritOperator(p.tok) == Lss ||
+		BinAritOperator(p.tok) == Gtr ||
+		BinAritOperator(p.tok) == Leq ||
+		BinAritOperator(p.tok) == Geq {
+		if compact && p.spaced {
+			return value
+		}
+		op := p.tok
+		pos := p.pos
+		p.nextArithOp(compact)
+		y := p.arithmExprShift(compact)
+		if y == nil {
+			p.followErrExp(pos, op.String())
+		}
+		value = &BinaryArithm{
+			OpPos: pos,
+			Op:    BinAritOperator(op),
+			X:     value,
+			Y:     y,
+		}
+	}
+	return value
+}
+
+func (p *Parser) arithmExprShift(compact bool) ArithmExpr {
+	value := p.arithmExpr3(compact)
+	for BinAritOperator(p.tok) == Shl ||
+		BinAritOperator(p.tok) == Shr {
+		if compact && p.spaced {
+			return value
+		}
+		op := p.tok
+		pos := p.pos
+		p.nextArithOp(compact)
+		y := p.arithmExpr3(compact)
+		if y == nil {
+			p.followErrExp(pos, op.String())
+		}
+		value = &BinaryArithm{
+			OpPos: pos,
+			Op:    BinAritOperator(op),
+			X:     value,
+			Y:     y,
+		}
+	}
+	return value
+}
+
+func (p *Parser) arithmExpr3(compact bool) ArithmExpr {
+	value := p.arithmExpr2(compact)
+	for BinAritOperator(p.tok) == Add ||
+		BinAritOperator(p.tok) == Sub {
+		if compact && p.spaced {
+			return value
+		}
+		op := p.tok
+		pos := p.pos
+		p.nextArithOp(compact)
+		y := p.arithmExpr2(compact)
+		if y == nil {
+			p.followErrExp(pos, op.String())
+		}
+		value = &BinaryArithm{
+			OpPos: pos,
+			Op:    BinAritOperator(op),
+			X:     value,
+			Y:     y,
+		}
+	}
+	return value
+}
+
+func (p *Parser) arithmExpr2(compact bool) ArithmExpr {
```

Comment:
A lot of these funcs copy the same ~20 lines and that's making me a bit uneasy. It's really easy to make mistakes when copying code around, or forgetting to update some of the copies when doing fixes.

Any thoughts on how we could deduplicate the code? I realise it would be a bit more inefficient to have one function with more branches, but honestly, I prefer a nearly-insignificant performance hit that brings an improvement in maintainability.

I also don't find the 1,2,3,4 function names super helpful :) I get that we also used numbers before, but to my mind those were isolated and didn't mean repeating code to this degree.

Reply:
I agree those names are super unhelpful, but I was inspired by bash: https://github.com/bminor/bash/blob/9e49d343e3cd7e20dad1b86ebfb764e8027596a7/expr.c#L780

I wanted to have the structure be identical to bash's, so the precedence rules are exactly the same. All the duplication is a bit weird, but I'm not sure how to improve it while keeping the same structure. I can try to improve it though, maybe I can add some helpers or something

Reply:
Here's an idea - a higher-level function that takes a list of operators to check against, and a function value as the "next" level. like:

```
type nextFunc func(compact bool, ops []token, next nextFunc) ArithmExpr

func arithmLevel(compact bool, ops []token, next nextFunc) ...
```

Of course, this will have a small perf hit, but I think it's worthwhile to avoid the ~5x copy-paste. An alternative would be to lay out these "levels" in a table, similar to what I do in master with the switch and level integers. Then, each table entry can tell you what operators need to be handled, and what the "next" level or function will be.

Reply:
In other words, I appreciate that recursive descent is easier to write, but it's also pretty hard to maintain when there are so many levels of precedence involved. This is why I coded it up with a table in master. Arguably, that was buggy, but I still think it should be possible to do it that way and not add another 200+ lines of mostly-repeated code.

Reply:
:+1: That's what I was working towards 😄 

I'll DRY it up

Reply:
I've refactored it, saved 200 lines of code! much better I think. Now all the functions look different at least